lib/Transforms/Speculation/SpeculationPlacement.cpp

```diff
@@ -0,0 +1,252 @@
+  this->saves.erase({srcOpResult, dstOp});
+}
+
+OpPlacement SpeculationPlacements::getSpeculatorPlacement() {
+  return this->speculator;
+}
+
+template <>
+const PlacementList &
+SpeculationPlacements::getPlacements<handshake::SpecSaveOp>() {
+  return this->saves;
+}
+
+template <>
+const PlacementList &
+SpeculationPlacements::getPlacements<handshake::SpecCommitOp>() {
+  return this->commits;
+}
+
+template <>
+const PlacementList &
+SpeculationPlacements::getPlacements<handshake::SpecSaveCommitOp>() {
+  return this->saveCommits;
+}
+
+inline void parseSpeculatorPlacement(
+    std::map<StringRef, llvm::SmallVector<PlacementNames>> &placements,
+    const llvm::json::Object *components) {
+  if (components->find("speculator") != components->end()) {
+    const llvm::json::Object *specObj = components->getObject("speculator");
+    StringRef srcOpName = specObj->getString("srcOp").value();
+    StringRef dstOpName = specObj->getString("dstOp").value();
+    placements["speculator"].push_back({srcOpName, dstOpName});
+  }
+}
+
+inline void parseOperationPlacements(
+    StringRef opType,
+    std::map<StringRef, llvm::SmallVector<PlacementNames>> &placements,
+    const llvm::json::Object *components) {
+  if (components->find(opType) != components->end()) {
+    const llvm::json::Array *jsonArray = components->getArray(opType);
+    for (const llvm::json::Value &element : *jsonArray) {
+      const llvm::json::Object *specObj = element.getAsObject();
+      StringRef srcOpName = specObj->getString("srcOp").value();
+      StringRef dstOpName = specObj->getString("dstOp").value();
+      placements[opType].push_back({srcOpName, dstOpName});
+    }
+  }
+}
+
+// JSON format example:
+// {
+//   "speculator": {"srcOp": "cmpi0", "dstOp":"fork5"},
+//   "saves": [
+//     {"srcOp": "extsi6", "dstOp":"mc_load0"},
+//     {"srcOp": "extsi7", "dstOp":"mc_load1"}
+//   ],
+//   "commits": [
+//     {"srcOp": "trunci0", "dstOp":"cond_br0"}
+//   ],
+//   "saveCommits": [
+//     {"srcOp": "mux1", "dstOp":"buffer10"}
+//   ]
+// }
+bool parseJSON(
+    const llvm::json::Value &jsonValue,
+    std::map<StringRef, llvm::SmallVector<PlacementNames>> &placements) {
+  const llvm::json::Object *components = jsonValue.getAsObject();
+  if (!components)
+    return false;
+
+  parseSpeculatorPlacement(placements, components);
+  parseOperationPlacements("saves", placements, components);
+  parseOperationPlacements("commits", placements, components);
+  parseOperationPlacements("save-commits", placements, components);
+  return true;
+}
+
+LogicalResult areOpConnected(Operation *srcOp, Operation *dstOp,
+                             Value &srcOpResult) {
+  unsigned result_idx = 0;
+  srcOpResult = srcOp->getResult(result_idx);
+  while (not llvm::is_contained(dstOp->getOperands(), srcOpResult)) {
+    result_idx++;
+    if (result_idx >= srcOp->getNumResults()) {
+      return failure(); // srcOp and dstOp are not connected
```

Comment:
This is an error in an end user's input file, so you should probably output the failure reason.